ext/xml/xml.c

```diff
@@ -1055,168 +1069,194 @@ PHP_FUNCTION(xml_set_object)
+		}
+	} else if (zend_parse_parameters_ex(ZEND_PARSE_PARAMS_QUIET, ZEND_NUM_ARGS(), "OSS", &pind, xml_parser_ce, &method_name, &method_name2) == SUCCESS) {
+		parser = Z_XMLPARSER_P(pind);
 
-	if (zend_parse_parameters(ZEND_NUM_ARGS(), "Oz", &pind, xml_parser_ce, &hdl) == FAILURE) {
-		RETURN_THROWS();
+		bool status = php_xml_check_string_method_arg(2, parser, method_name, &start_fcc);
+		if (status == false) {
+			RETURN_FALSE;
+		}
+		bool status2 = php_xml_check_string_method_arg(3, parser, method_name2, &end_fcc);
+		if (status2 == false) {
+			RETURN_FALSE;
+		}
+	} else {
+		zval *dummy_start;
+		zval *dummy_end;
+		if (zend_parse_parameters(ZEND_NUM_ARGS(), "Ozz", &pind, xml_parser_ce, &dummy_start, &dummy_end) == FAILURE) {
+			RETURN_THROWS();
+		} else {
+			switch (Z_TYPE_P(dummy_start)) {
+				case IS_NULL:
+				case IS_STRING:
+					break;
+				default:
+					zend_argument_type_error(2, "must be of type callable|string|null");
+					RETURN_THROWS();
+			}
+			zend_argument_type_error(3, "must be of type callable|string|null");
+			RETURN_THROWS();
+		}
 	}
 
-	parser = Z_XMLPARSER_P(pind);
-	xml_set_handler(&parser->externalEntityRefHandler, hdl);
-	XML_SetExternalEntityRefHandler(parser->parser, (void *) _xml_externalEntityRefHandler);
+	xml_set_handler(&parser->startElementHandler, &start_fcc);
+	xml_set_handler(&parser->endElementHandler, &end_fcc);
+	XML_SetElementHandler(parser->parser, _xml_startElementHandler, _xml_endElementHandler);
 
 	RETURN_TRUE;
 }
 /* }}} */
 
-/* {{{ Set up character data handler */
-PHP_FUNCTION(xml_set_start_namespace_decl_handler)
-{
-	xml_parser *parser;
-	zval *pind, *hdl;
-
-	if (zend_parse_parameters(ZEND_NUM_ARGS(), "Oz", &pind, xml_parser_ce, &hdl) == FAILURE) {
+static void php_xml_set_handler_parse_callable(
+	INTERNAL_FUNCTION_PARAMETERS,
+	xml_parser **const parser,
+	zend_fcall_info_cache *const parser_handler_fcc
+) {
+	zval *pind;
+	zend_fcall_info handler_fci = {0};
+	zend_fcall_info_cache handler_fcc = {0};
+	zend_string *method_name = NULL;
+
+	if (zend_parse_parameters_ex(ZEND_PARSE_PARAMS_QUIET, ZEND_NUM_ARGS(), "Of!", &pind, xml_parser_ce, &handler_fci, &handler_fcc) == SUCCESS) {
+		*parser = Z_XMLPARSER_P(pind);
+		if (!ZEND_FCI_INITIALIZED(handler_fci)) {
+			/* Free handler, so just return and a uninitialized FCC communicates this */
+			return;
+		}
+		if (!ZEND_FCC_INITIALIZED(handler_fcc)) {
+			zend_is_callable_ex(&handler_fci.function_name, NULL, IS_CALLABLE_SUPPRESS_DEPRECATIONS, NULL, &handler_fcc, NULL);
+			/* Call trampoline has been cleared by zpp. Refetch it, because we want to deal
+			 * with it ourselves. It is important that it is not refetched on every call,
+			 * because calls may occur from different scopes. */
+		}
+		memcpy(parser_handler_fcc, &handler_fcc, sizeof(zend_fcall_info_cache));
+	} else if (zend_parse_parameters_ex(ZEND_PARSE_PARAMS_QUIET, ZEND_NUM_ARGS(), "OS", &pind, xml_parser_ce, &method_name) == SUCCESS) {
+		xml_parser *local_parser;
```

Comment:
Why do we need the `local_parser` variable?

Reply:
The compiler realises that it doesn't need to refetch `*parser` directly after writing it, it "forwards" the store.